### ISSUE TITLE
Neovim 0.9.1 + nvim-treesitter hangs when changing Git branches/refs

### ISSUE THREAD

**Issue by itaranto** [neovim/neovim#24186]
### Problem

Neovim 0.9.1 hangs when `nvim-treesitter` is enabled, this happens (presumably) after moving/re-creating files while having open buffers.

This happens 100% of the time when switching Git branches.

I tested `nvim-treesitter` with both `0.9.0 ` and the lastest `master` commit too (`bad9f5bf5b4af5d97a9cffdc957f1b23dc91d851`).

This is my config (lazy.nvim):

```lua
return {
  'nvim-treesitter/nvim-treesitter',
  version = '*',
  build = ':TSUpdate',
  config = function()
    require('nvim-treesitter.configs').setup({
      ensure_installed = 'all',
      highlight = { enable = true },
    })

    vim.opt.foldmethod = 'expr'
    vim.opt.foldexpr = 'v:lua.vim.treesitter.foldexpr()'
  end,
}
```

I'm pretty sure `nvim-treesitter` is causing this since the issue is no longer reproducible after disabling it.

I also tested with no plugins except `nvim-treesitter` and the problem is present.

Neovim nightly (`v0.10.0-dev-591+g42f9573e5d`)  nor `0.8.0` doesn't seem to have this issue.

I initially reported this to [nvim-treesitter](https://github.com/nvim-treesitter/nvim-treesitter/issues/5027).

### Steps to reproduce

1. Go to some Git repository
2. Open a file with Neovim
3. While having that file open in a buffer, switch to another Git branch or ref
4. Neovim hangs and its completely unresponsive. 
   After a few seconds its CPU usage goes to 100%. It can be killed with `SIGTERM` (`pkill nvim`) .

### Expected behavior

Neovim does not hang.

### Neovim version (nvim -v)

0.9.1

### Vim (not Nvim) behaves the same?

I don't have Vim.

### Operating system/version

Arch Linux

### Terminal name/version

Alacritty 0.12.1

### $TERM environment variable

xterm-256color

### Installation

Arch official repositories

**Comment by clason** [comment]
Please try to reproduce without nvim-treesitter; just install the (needed!) parser and call `vim.treesitter.start()` in a `Filetype` autocommand.

**Comment by itaranto** [comment]
> Please try to reproduce without nvim-treesitter; just install the (needed!) parser and call `vim.treesitter.start()` in a `Filetype` autocommand.
> 
> I can't reproduce it with just the provided "steps". We need a minimal config to test with `nvim --clean -u minimal.lua <file>`.

Well, I have some news, the problem seems to be this part:

```lua
    vim.opt.foldmethod = 'expr'
    vim.opt.foldexpr = 'v:lua.vim.treesitter.foldexpr()'
 ```

If I comment that part, the issue goes away.

I've been able to reproduce it 100% of the time like this:

- Open a `.go` file in a buffer. Make sure this file is modified in two different Git branches: A and B.
- In another terminal, switch from branch A to branch B.
Result: Neovim hangs.

Here's the minimal.lua file that I used with `nvim --clean -u minimal.lua`:

```lua
vim.opt.foldmethod = 'expr'
vim.opt.foldexpr = 'v:lua.vim.treesitter.foldexpr()'

vim.treesitter.language.add('go', { path = '/home/ignacio/.local/share/nvim/lazy/nvim-treesitter/parser/go.so' })

vim.api.nvim_create_autocmd('FileType', {
  pattern = 'go',
  callback = function(args)

    vim.treesitter.start(args.buf, 'go')
    vim.bo[args.buf].syntax = 'on'
  end,
})
```lua

**Comment by clason** [comment]
>  vim.bo[args.buf].syntax = 'on'

Is this part necessary? You should normally not need it.

**Comment by clason** [comment]
Wait, I completely missed the part that master is fine. Then the problem is already fixed.

**Comment by itaranto** [comment]
> Is this part necessary? You should normally not need it.

No, it's not. I just added it because it was part of the example present in `:help vim.treesitter.start()`.

> Wait, I completely missed the part that master is fine. Then the problem is already fixed.

Yes, but still, don't you want to know what's the root cause so it doesn't happen again?

**Comment by itaranto** [comment]
> Not really lol. We can use that time to fix on the countless bugs that still exists instead. If you want to bisect the commit where this was fixed then go for it.

Fair enough. I may do the bisect when I have some free time :smiley:

**Comment by itaranto** [comment]
It seems this issue is still present in `0.9.2`.

**Comment by freddiehaddad** [comment]
I'm hitting this error as well. Switching git branches causes this problem. Removing the fold expression gets rid of it for me.

**Comment by freddiehaddad** [comment]
Confident this is a treesitter issue.  I can reproduce it every time with the following steps:

1. Modify a git tracked file inside neovim and save it.  Leave the buffer open and visible.
2. From a separate terminal window, do a git restore of that file.
3. Go back to neovim and it will hang.

lazy.nvim plugin spec (enough to make it happen consistently):

```lua
{
	'nvim-treesitter/nvim-treesitter',
	build = ':TSUpdate',
	init = function()
		vim.opt.foldenable = false
		vim.opt.foldmethod = 'expr'
		vim.opt.foldexpr = 'v:lua.vim.treesitter.foldexpr()'
	end,
	main = 'nvim-treesitter.configs',
	opts = {
		auto_install = true,
		highlight = {
			enable = true,
		},
		incremental_selection = {
			enable = true,
			keymaps = {
				init_selection = '<c-i>',
				node_incremental = '<c-i>',
				scope_incremental = false,
				node_decremental = '<bs>',
			},
		},
		indent = {
			enable = true,
		},
	},
}
```

**Comment by freddiehaddad** [comment]
Since this issue is closed, I opened: (#25608).
